File: desafio4/go/io.go
package main

import ("os"; "log"; "io/ioutil"; "bufio"; "fmt")

type BitInputStream struct {
	buffer uint16
	pos  int
	bitsInBuffer  uint16
	eof  bool
	bytes  []byte
	len  int
}

type BitOutputStream struct {
	out *bufio.Writer
	buffer uint16
	bitsInBuffer uint16
}

func NewBitInputStream(filename string) *BitInputStream {
	p := new(BitInputStream)
	p.buffer = 0
	p.pos = 0
	p.bitsInBuffer = 0
	p.eof = false

	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	bytes, err := ioutil.ReadAll(file)
	if err != nil {
		log.Fatal(err)
	}
	p.bytes = bytes
	p.len = len(bytes)
	p.fillBuffer()
	return p
}

func NewBitOutputStream(filename string) *BitOutputStream {
	p := new(BitOutputStream)
	file, err := os.Create(filename)
	if err != nil {
		log.Fatal(err)
	}
	p.out = bufio.NewWriter(file)
	p.buffer = 0
	p.bitsInBuffer = 0
	return p
}

func (i *BitInputStream) ReadBool() bool {
	if i.eof { log.Fatal("can't read bool from input stream") }

	i.bitsInBuffer -= 1

	bit := ((i.buffer >> i.bitsInBuffer) & 1) == 1
	if i.bitsInBuffer == 0 {
		i.fillBuffer()
	}
	return bit
}

func (i *BitInputStream) ReadChar() byte {
	if i.eof { log.Fatal("reading from empty input stream") }

	fmt.Printf("read char pos = %d\n", i.pos)

	x := i.buffer
	if i.bitsInBuffer == 8 {
		i.fillBuffer()
	} else {
		x = x << (8 - i.bitsInBuffer)
		temp := i.bitsInBuffer
		i.fillBuffer()
		i.bitsInBuffer = temp
		x |= i.buffer >> i.bitsInBuffer
	}
	return byte(x & 0xFF)
}

func (i *BitInputStream) ReadInt() uint32 {
	fmt.Printf("read int pos = %d\n", i.pos)
	x := uint32(i.ReadChar())
	x = (x << 8) | uint32(i.ReadChar())
	x = (x << 8) | uint32(i.ReadChar())
	x = (x << 8) | uint32(i.ReadChar())
	return x
}

func (i *BitInputStream) GetBytes() []byte {
	return i.bytes
}

func (i *BitInputStream) fillBuffer() {
	if i.pos == i.len {
		i.eof = true
	} else {
		i.buffer = uint16(i.bytes[i.pos])
		i.bitsInBuffer = 8
		i.pos++
	}
}

func (o *BitOutputStream) clearBuffer() {
	if o.bitsInBuffer == 0 { return }
	if o.bitsInBuffer > 0 {
		o.buffer = o.buffer << (8 - o.bitsInBuffer)
	}
	o.out.WriteByte(byte(o.buffer))
	o.buffer = 0
	o.bitsInBuffer = 0
}

func (o *BitOutputStream) WriteBit(bit uint16) {
	if bit != 0 && bit != 1 { log.Fatal("argument must be 0 or 1.") }
	o.buffer = (o.buffer << 1) | bit
	o.bitsInBuffer++
	if o.bitsInBuffer == 8 {
		o.clearBuffer()
	}
}

func (o *BitOutputStream) WriteByte(b uint16) {
	for i := 0; i < 8; i++ {
		bit := (b >> (8 - uint16(i) - 1)) & 1
		o.WriteBit(bit)
	}
}

func (o *BitOutputStream) WriteInt(i uint32) {
	o.WriteByte(uint16((i >> 24) & 0xFF))
	o.WriteByte(uint16((i >> 16) & 0xFF))
	o.WriteByte(uint16((i >>  8) & 0xFF))
	o.WriteByte(uint16((i >>  0) & 0xFF))
}

func (o *BitOutputStream) Close() {
	o.clearBuffer()
	o.out.Flush()
}

File: desafio4/swift/Sources/io.swift
import Foundation

func _panic(msg : String) {
	print(msg)
	exit(-1)
}

class BitInputStream {
	var buffer : UInt16 = 0
	var pos : Int = 0
	var bitsInBuffer : UInt16 = 0
	var eof : Bool = false
	var bytes : [UInt8] = [UInt8]()
	var len : Int = 0

	init(inputFile: String) {
		if let data = NSData(contentsOfFile: inputFile) {
			bytes = [UInt8](repeating: 0, count: data.length)
			data.getBytes(&bytes, length: data.length)
			len = data.length
		}
		fillBuffer()
	}

	func readBool() -> Bool {
		if eof {
			_panic(msg: "can't read bool from input stream")
		}

		bitsInBuffer -= 1

		let bit = ((buffer >> bitsInBuffer) & 1) == 1
		if bitsInBuffer == 0 {
			fillBuffer()
		}
		return bit
	}

	func readChar() -> UInt8 {
		if eof {
			_panic(msg: "reading from empty input stream")
		}

		var x = buffer
		if bitsInBuffer == 8 {
			fillBuffer()
		} else {
			x = x << (8 - bitsInBuffer)
			let temp = bitsInBuffer
			fillBuffer()
			bitsInBuffer = temp
			x |= buffer >> bitsInBuffer
		}
		return UInt8(x & 0xFF)
	}

	func readInt() -> UInt32 {
		var x = UInt32(readChar())
		x = (x << 8) | UInt32(readChar())
		x = (x << 8) | UInt32(readChar())
		x = (x << 8) | UInt32(readChar())
		return x
	}

	func fillBuffer() {
		if pos == len {
			eof = true
		} else {
			buffer = UInt16(bytes[pos])
			bitsInBuffer = 8
			pos += 1
		}
	}

	func getBytes() -> [UInt8] {
		return bytes
	}
}

class BitOutputStream {
	var out : Data

	var buffer : UInt16 = 0
	var bitsInBuffer : UInt16 = 0
	var output : String

	init(outputFile: String) {
		out = Data()
		output = outputFile
	}

	func writeBitc(bit: Character) {
		if bit == "0" {
			writeBit(bit: 0)
		} else if bit == "1" {
			writeBit(bit: 1)
		} else {
			_panic(msg: "argument must be '0' or '1'")
		}
	}

	func writeBit(bit: UInt16) {
		if bit != 0 && bit != 1 {
			_panic(msg: "argument must be 0 or 1")
		}
		buffer = (buffer << 1) | bit
		bitsInBuffer += 1
		if bitsInBuffer == 8 {
			clearBuffer()
		}
	}

	func writeByte(byte: UInt16) {
		for i in 0..<8 {
			let bit = (byte >> (8 - i - 1)) & 1
			writeBit(bit: bit)
		}
	}

	func writeInt(i: UInt32) {
		writeByte(byte: UInt16((i >> 24) & 0xFF))
		writeByte(byte: UInt16((i >> 16) & 0xFF))
		writeByte(byte: UInt16((i >>  8) & 0xFF))
		writeByte(byte: UInt16((i >>  0) & 0xFF))

	}

	func clearBuffer() {
		if bitsInBuffer == 0 { return }
		if bitsInBuffer > 0 {
			buffer = buffer << (8 - bitsInBuffer)
		}
		out.append(UInt8(buffer))
		buffer = 0
		bitsInBuffer = 0
	}

	func close() {
		let nsdata = NSData(data: out)
		nsdata.write(toFile: output, atomically: true)
	}
}